qemu/tests/watchdog.py

```diff
@@ -168,8 +169,12 @@ def check_watchdog_support():
         """
         qemu_binary = utils_misc.get_qemu_binary(params)
 
-        watchdog_type_check = params.get(
-            "watchdog_type_check", " -watchdog '?'")
+        if "aarch" in platform.machine():
+            machine_type = params.get("machine_type").split(':', 1)[1]
+            watchdog_arg = " -M %s -watchdog '?'" % machine_type
+        else:
+            watchdog_arg = " -watchdog '?'"
+        watchdog_type_check = params.get("watchdog_type_check", watchdog_arg)
```

Comment:
I am thinking that, if you specify `watchdog_arg` here, then the definition of the parameter `watchdog_type_check` is meaningless. Either keep `watchdog_type_check` and define it in 'watchdog.cfg' or it can be removed, and I would prefer the first approach.

Reply:
I have same idea with @vivianQizhu, the two params are repetitive in a sense.  
And I am thinking about appoint machine type for all, although it is not must for some platform. This means we can define `watchdog_type_check = '-M %s -watchdog ?'` in cfg for all, and update it in script,  Could we ?

Reply:
How about?
1. list all supported machines on host
2. list all supported watchdog on all machines
3. check the host support watchdog types
```
    def check_watchdog_support():
        """
        check the host qemu-kvm support watchdog device
        Test Step:
        1. Send qemu command 'qemu -watchdog ?'
        2. Check the watchdog type that the host support.
        """
        qemu_binary = utils_misc.get_qemu_binary(params)

        # list all supported machines on host
        list_machines_cmd = qemu_binary + ' -M ? | awk \'NR>2{print p}{p=$1}\''
        machine_types = process.system_output(
                list_machines_cmd, shell=True).decode().splitlines()

        # list all supported watchdog on all machines
        watchdog_devices = []
        for machine in machine_types:
            qemu_cmd = qemu_binary + " -M %s -watchdog '?'" % machine
            watchdog_devices.append(process.system_output(
                    "%s 2>&1" % qemu_cmd, shell=True).decode())

        # check the host support watchdog types.
        if watchdog_devices:
            for watchdog_device in watchdog_devices:
                error_context.context(
                    "Checking whether or not the host support"
                    " WDT '%s'" % watchdog_device_type, logging.info)
                if re.findall(watchdog_device_type, watchdog_device, re.I):
                    logging.info("The host support '%s' type watchdog device" %
                                 watchdog_device_type)
                    break
            else:
                logging.info("The host support watchdog device type is: '%s'"
                             % watchdog_devices)
                test.cancel("watdog %s isn't supported" % watchdog_device_type)
        else:
            test.cancel("No watchdog device supported by the host!")
```

FYI:
```
(workspace) [root@hp-dl388g8-16 kar]#  /usr/libexec/qemu-kvm -M ? | awk 'NR>2{print p}{p=$1}'
pc
pc-i440fx-rhel7.6.0
pc-i440fx-rhel7.5.0
pc-i440fx-rhel7.4.0
pc-i440fx-rhel7.3.0
pc-i440fx-rhel7.2.0
pc-i440fx-rhel7.1.0
pc-i440fx-rhel7.0.0
q35
pc-q35-rhel8.3.0
pc-q35-rhel8.2.0
pc-q35-rhel8.1.0
pc-q35-rhel8.0.0
pc-q35-rhel7.6.0
pc-q35-rhel7.5.0
pc-q35-rhel7.4.0
pc-q35-rhel7.3.0

(workspace) [root@hp-dl388g8-16 kar]# /usr/libexec/qemu-kvm -M pc-q35-rhel7.3.0 -watchdog '?'
	ib700	iBASE 700
	i6300esb	Intel 6300ESB
(workspace) [root@hp-dl388g8-16 kar]# /usr/libexec/qemu-kvm -M pc -watchdog '?'
	ib700	iBASE 700
	i6300esb	Intel 6300ESB
(workspace) [root@hp-dl388g8-16 kar]# /usr/libexec/qemu-kvm -M q35 -watchdog '?'
	ib700	iBASE 700
	i6300esb	Intel 6300ESB
(workspace) [root@hp-dl388g8-16 kar]# /usr/libexec/qemu-kvm -M none -watchdog '?'
	ib700	iBASE 700
	i6300esb	Intel 6300ESB


```

Reply:
@xiangchunfu Do you have any comments here? Thanks.

Reply:
> 2\. list all supported watchdog on all machines
Hi, all,  
The aarch64 platform does not have a default machine value. 
The watchdog device  can be queried only after the specified machine type. 
(cmd eg: /usr/libexec/qemu-kvm -M virt -watchdog '?')

But，the x86 platform has a default value, which does not need to be specified. 
In QEMU 5.0 version , the default machine value is pc-i440fx-5.0. 
```
[root@localhost /]# /usr/libexec/qemu-kvm -M ?
Supported machines are:
microvm              microvm (i386)
pc                   Standard PC (i440FX + PIIX, 1996) (alias of pc-i440fx-5.0)
_pc-i440fx-5.0        Standard PC (i440FX + PIIX, 1996) (default)_
pc-i440fx-4.2        Standard PC (i440FX + PIIX, 1996)
```

So,  the difference is that aarch64 has a default value and the x86 does not have a default value. And this test case needs to be adapted for it. 

I don't think we need to test all machine types. 
In fact, we can find that on the x86 platforms, all machine-supported watchdog devices are the same. The same result is true for the aarch64 platform.

Reply:
On the x86 platform, all machines support the watchdog：
 ib700   iBASE 700
 i6300esb        Intel 6300ESB

On the aarch64 platform, all machines support the watchdog：
 aspeed.wdt      Aspeed watchdog device
 i6300esb        Intel 6300ESB
 imx2-watchdog   i.MX2 Watchdog

Reply:
> On the x86 platform, all machines support the watchdog：
> ib700 iBASE 700
> i6300esb Intel 6300ESB
> 
> On the aarch64 platform, all machines support the watchdog：
> aspeed.wdt Aspeed watchdog device
> i6300esb Intel 6300ESB
> imx2-watchdog i.MX2 Watchdog

Hi @Kuhn-Chen 
so the watchdog devices from the command `/usr/libexec/qemu-kvm -M virt -watchdog '?'`  is not included in the above list?

Reply:
> > On the x86 platform, all machines support the watchdog：
> > ib700 iBASE 700
> > i6300esb Intel 6300ESB
> > On the aarch64 platform, all machines support the watchdog：
> > aspeed.wdt Aspeed watchdog device
> > i6300esb Intel 6300ESB
> > imx2-watchdog i.MX2 Watchdog
> 
> Hi @Kuhn-Chen
> so the watchdog devices from the command `/usr/libexec/qemu-kvm -M virt -watchdog '?'` is not included in the above list?

No，all machines have the same result：
```
[root@target /]# /usr/libexec/qemu-kvm -M virt -watchdog ?
        aspeed.wdt      Aspeed watchdog device
        i6300esb        Intel 6300ESB
        imx2-watchdog   i.MX2 Watchdog
```

Reply:
Hi @YongxueHong @Kuhn-Chen Let's just focus on the original issue. To test all supported devices actually means a test requirement change, which should be initiated by the feature owner. For this patch let's just focus one the aarch platform support. Thanks.

Reply:
Hi, @vivianQizhu 
Okay, please ignore me.
Thanks.

Reply:
I suggest adding `watchdog_type_check = -M virt -watchdog help` for aarch64 platform, and virt board is the best choice. https://qemu.readthedocs.io/en/latest/system/target-arm.html#choosing-a-board-model
@Kuhn-Chen In fact, watchdog is not supported for all distributions.
Upstream:
```shell
$ ./aarch64-softmmu/qemu-system-aarch64 -M virt -watchdog help
	imx2-watchdog	i.MX2 Watchdog
	aspeed.wdt	Aspeed watchdog device
	i6300esb	Intel 6300ESB
$
```
RHEL:
```shell
$ /usr/libexec/qemu-kvm -M virt -watchdog help
$
```

Reply:
> I suggest adding `watchdog_type_check = -M virt -watchdog help` for aarch64 platform, and virt board is the best choice. https://qemu.readthedocs.io/en/latest/system/target-arm.html#choosing-a-board-model
> @Kuhn-Chen In fact, watchdog is not supported for all distributions.
> Upstream:
> 
> ```shell
> $ ./aarch64-softmmu/qemu-system-aarch64 -M virt -watchdog help
> 	imx2-watchdog	i.MX2 Watchdog
> 	aspeed.wdt	Aspeed watchdog device
> 	i6300esb	Intel 6300ESB
> $
> ```
> 
> RHEL:
> 
> ```shell
> $ /usr/libexec/qemu-kvm -M virt -watchdog help
> $
> ```

So, let's give it a default machine for aarch64 platform, and this looks like a good idea. 
Hi，@vivianQizhu @yanan-fu @YongxueHong @PaulYuuu , do you have any other suggestions? 
The code is as follows: 
```
        if "aarch64" in platform.machine():
            watchdog_type_check = params.get("watchdog_type_check",
                    " -M virt -watchdog '?'")
        else:
            watchdog_type_check = params.get("watchdog_type_check", 
                    " -watchdog '?'")
```

Reply:
A better workaround is:
qemu/tests/cfg/watchdog.cfg
```diff
--- a/qemu/tests/cfg/watchdog.cfg
+++ b/qemu/tests/cfg/watchdog.cfg
@@ -3,6 +3,8 @@
     enable_watchdog = yes
     start_vm = no
     type = watchdog
+    aarch64:
+        watchdog_type_check = " -M virt -watchdog help"
     variants:
         -i6300esb:
             no RHEL.4

Reply:
> A better workaround is:
> qemu/tests/cfg/watchdog.cfg
> 
> ```diff
> --- a/qemu/tests/cfg/watchdog.cfg
> +++ b/qemu/tests/cfg/watchdog.cfg
> @@ -3,6 +3,8 @@
>      enable_watchdog = yes
>      start_vm = no
>      type = watchdog
> +    aarch64:
> +        watchdog_type_check = " -M virt -watchdog help"
>      variants:
>          -i6300esb:
>              no RHEL.4
> ```
Oh, Yes you are right！ I haven't been in the habit of modifying the cfg.

Reply:
I agree with @PaulYuuu To edit cfg file is sufficient, that's why we define `watchdog_type_check` in Params.

Reply:
> I agree with @PaulYuuu To edit cfg file is sufficient, that's why we define `watchdog_type_check` in Params.
Yes ，I agress too！

Reply:
> > I suggest adding `watchdog_type_check = -M virt -watchdog help` for aarch64 platform, and virt board is the best choice. https://qemu.readthedocs.io/en/latest/system/target-arm.html#choosing-a-board-model
> > @Kuhn-Chen In fact, watchdog is not supported for all distributions.
> > Upstream:
> > ```shell
> > $ ./aarch64-softmmu/qemu-system-aarch64 -M virt -watchdog help
> > 	imx2-watchdog	i.MX2 Watchdog
> > 	aspeed.wdt	Aspeed watchdog device
> > 	i6300esb	Intel 6300ESB
> > $
> > ```
> > 
> > 
> > RHEL:
> > ```shell
> > $ /usr/libexec/qemu-kvm -M virt -watchdog help
> > $
> > ```
> 
> So, let's give it a default machine for aarch64 platform, and this looks like a good idea.
> Hi，@vivianQizhu @yanan-fu @YongxueHong @PaulYuuu , do you have any other suggestions?
> The code is as follows:
> 
> ```
>         if "aarch64" in platform.machine():
>             watchdog_type_check = params.get("watchdog_type_check",
>                     " -M virt -watchdog '?'")
>         else:
>             watchdog_type_check = params.get("watchdog_type_check", 
>                     " -watchdog '?'")
> ```

> > I suggest adding `watchdog_type_check = -M virt -watchdog help` for aarch64 platform, and virt board is the best choice. https://qemu.readthedocs.io/en/latest/system/target-arm.html#choosing-a-board-model

The virt has been discussed as a default machine value in qemu community.
https://lists.nongnu.org/archive/html/qemu-devel/2019-09/msg03627.html
Unfortunately, they did not reach a consensus.  But we use virt  here should be reasonable!

Reply:
> I agree with @PaulYuuu To edit cfg file is sufficient, that's why we define `watchdog_type_check` in Params.
Hi ，@vivianQizhu  @PaulYuuu ，The code has been updated. Please review it. 
Thanks you！

Reply:
The test result of a test case is as follows: 

 avocado run watchdog.heartbeat_test.invalid_excp.i6300esb --vt-type qemu --vt-guest-os Guest.Linux.Fedora.28 --vt-machine-type arm64-pci
JOB ID     : 9f050b07bcaf16a78e57ebd7e84ba133a728cfb2
JOB LOG    : /root/avocado/job-results/job-2020-08-05T14.35-9f050b0/job.log
 (1/1) type_specific.io-github-autotest-qemu.watchdog.heartbeat_test.invalid_excp.i6300esb: PASS (35.89 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 38.81 s


```
2020-08-05 14:35:15,242 test             L0318 DEBUG|     vm_type = qemu
2020-08-05 14:35:15,242 test             L0318 DEBUG|     vms = avocado-vt-vm1
2020-08-05 14:35:15,242 test             L0318 DEBUG|     vnc_autoport = yes
2020-08-05 14:35:15,242 test             L0318 DEBUG|     watchdog_action = pause
2020-08-05 14:35:15,243 test             L0318 DEBUG|     watchdog_device_type = i6300esb
2020-08-05 14:35:15,243 test             L0318 DEBUG|     watchdog_type_check =  -M virt -watchdog ?
2020-08-05 14:35:15,248 test             L0372 DEBUG| Searching for test modules that match 'type = watchdog' and 'provider = io-github-autotest-qemu' on this cartesian dict
2020-08-05 14:35:15,249 test             L0386 DEBUG| Found subtest module /var/lib/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/watchdog.py
2020-08-05 14:35:15,255 utils_env        L0082 WARNI| Creating new, empty env file
2020-08-05 14:35:16,383 env_process      L1061 DEBUG| KVM version: 4.19.90-2005.1.0.0038.oe1.aarch64
2020-08-05 14:35:16,444 env_process      L1093 DEBUG| KVM userspace version(qemu): 4.1.0 (qemu-4.1.0-13.oe1)
2020-08-05 14:35:16,463 storage          L0722 INFO | backup image size: 1536950272, available size: 33125920768.
2020-08-05 14:35:16,463 storage          L0737 DEBUG| Image backup /var/lib/avocado/data/avocado-vt/images/f28-aarch64.qcow2.backup already exists, skipping...
2020-08-05 14:35:16,481 env_process      L1925 WARNI| registers is not alive. Can't query the avocado-vt-vm1 status
2020-08-05 14:35:16,482 error_context    L0079 INFO | Context: Checking whether or not the host support WDT 'i6300esb'
2020-08-05 14:35:16,482 process          L0604 INFO | Running '/usr/libexec/qemu-kvm -M virt -watchdog ? 2>&1'
2020-08-05 14:35:16,544 process          L0436 DEBUG| [stdout] Unable to init server: Could not connect: Connection refused
2020-08-05 14:35:16,545 process          L0436 DEBUG| [stdout]  aspeed.wdt      Aspeed watchdog device
2020-08-05 14:35:16,546 process          L0436 DEBUG| [stdout]  i6300esb        Intel 6300ESB
2020-08-05 14:35:16,546 process          L0436 DEBUG| [stdout]  imx2-watchdog   i.MX2 Watchdog
2020-08-05 14:35:16,547 process          L0686 INFO | Command '/usr/libexec/qemu-kvm -M virt -watchdog ? 2>&1' finished with 0 after 0.06337332725524902s
2020-08-05 14:35:16,548 watchdog         L0183 INFO | The host support 'i6300esb' type watchdog device
2020-08-05 14:35:16,548 error_context    L0079 INFO | Context: 'heartbeat_test' test starting ..
```

Reply:
> I have same idea with @vivianQizhu, the two params are repetitive in a sense.
> And I am thinking about appoint machine type for all, although it is not must for some platform. This means we can define `watchdog_type_check = '-M %s -watchdog ?'` in cfg for all, and update it in script, Could we ?

I would like to clarify the meaning of `all` in the comment above ~ 
It does not means traverse all the supported machine type.   Just means use machine type of the current test(in params) to `watchdog_type_check` while is it not must for x86, and also ppc i think.

It is not must,  so the current change is ok for me too ~   thanks!